step-release-vis/src/app/services/environment.ts

```diff
@@ -15,33 +14,38 @@ export class EnvironmentService {
   }
 
   // xs: 0-100, ys: timestamps
-  // TODO(naoai): write test
   getPolygons(jsonFile: string): Observable<Polygon[]> {
     return this.readJson(jsonFile)
       .pipe(map(environments => this.calculatePolygons(environments)));
   }
 
+  // TODO(naoai): compute the coordinates for the polygons
   private calculatePolygons(environments: Environment[]): Polygon[] {
     return [];
   }
 
-  private extractPoints(candsInfo: CandidateInfo[]): CandidatesSnapshot {
-    return [];
-  }
-
-  // TODO(ancar): calculate the percentages for cands at given timestamp
-  private getPercentages(candsInfo: CandidateInfo[]): CandidatesSnapshot {
-    return [];
+  private getPercentages(candsInfo: CandidateInfo[]): Map<string, number> {
+    const candInfo2percentage: Map<string, number> = new Map();
+    let totalJobSum = 0;
+
+    for (const candInfo of candsInfo) {
+      totalJobSum += candInfo.job_count;
+    }
+
+    let percentageSum = 0;
+    for (const candInfo of candsInfo) {
+      const percentage = Math.floor(candInfo.job_count / totalJobSum * 100);
+      candInfo2percentage.set(candInfo.name, percentage);
+      percentageSum += percentage;
+    }
+
+    // treat rounding error here
+    const prevValue = candInfo2percentage.get(candsInfo[0].name);
+    candInfo2percentage.set(candsInfo[0].name, prevValue + 100 - percentageSum);
```

Comment:
What is this part for? Should that be really candsInfo[0] on every run?

Reply:
it does not really matter. since we do not want percentages to be 33.33% exactly for the representation, i rounded all down and added the difference to the first one so the sum is 100 and we do not have precision errors.